R/deploy-site.R

```diff
@@ -0,0 +1,84 @@
+#' Deploy a pkgbown site on Travis-CI using Github Pages
+#'
+#' @inheritParams build_site
+#' @param tarball The location of the built package tarball. The default Travis
+#'   configuration for R packages sets `PKG_TARBALL` to this path.
+#' @param ssh_id The private id to use, a base64 encoded value, by default this
+#'   is read from the id_rsa environment variable. It is recommended you add this
+#'   key as a private environment variable in the Travis repository settings.
+#' @param ssh_id_file The file location to write the ssh private key obtained from `ssh_id`.
+#' @param repo_slug The `user/repo` slug for the repository.
+#' @param commit_message The commit message to be used for the commit.
+#' @param verbose Print verbose output
+#' @param ... Additional arguments passed to [build_site()].
+#' @export
+deploy_site <- function(
+  pkg = ".",
+  tarball = Sys.getenv("PKG_TARBALL", ""),
+  ssh_id = Sys.getenv("id_rsa", ""),
+  ssh_id_file = "~/.ssh/id_rsa",
+  repo_slug = Sys.getenv("TRAVIS_REPO_SLUG", ""),
+  commit_message = "build site",
+  verbose = FALSE,
+  ...) {
+
+  if (!nzchar(tarball)) {
+    stop("No built tarball detected, please provide the location of one with `tarball`", call. = FALSE)
+  }
+  rule("Deploying site", line = 2)
+  rule("Installing package", line = 1)
+  callr::rcmd("INSTALL", tarball, show = verbose, fail_on_status = TRUE)
+
+  build_site(pkg, ...)
+
+  if (!nzchar(ssh_id)) {
+    stop("Need to setup a deploy key with `ssh-keygen`, base64 encode it and then assign it to `id_rsa`", call. = FALSE)
+  }
+
+  if (!nzchar(repo_slug)) {
+    stop("No repo detected, please supply one with `repo_slug`", call. = FALSE)
+  }
+
+  remote_url <- glue::glue("git@github.com:{repo_slug}.git")
+
+
+  rule("Setting up SSH id", line = 1)
+  cat_line("Copying private key to: ", ssh_id_file)
+  write_lines(rawToChar(openssl::base64_decode(ssh_id)), ssh_id_file)
+  cat_line("Setting private key permissions", ssh_id_file)
+  fs::file_chmod(ssh_id_file, "0600")
+
+
+  rule("Commiting built site", line = 1)
+  withr::with_dir("docs", {
+    processx::run("git", "init", echo = verbose)
+    processx::run("git", c("add", "-A", "."), echo = verbose)
+    processx::run("git", c("remote", "add", "origin", remote_url), echo = verbose)
+    processx::run("git", c("commit", "-m", commit_message), echo = verbose)
```

Comment:
This code always clobbers the history in the gh-pages branch, which I think is actually preferred, often the data in the built site is binary (such as images) and doesn't diff well.

If desired we could support simply appending a new commit as well however.

Reply:
I think it's fine to clobber, although I'd like to see some safety checks to make sure we don't clobber the wrong thing. Maybe check that the history is short?

Reply:
Maybe we could check that directory is either empty (for first run) or contains some pkgdown file that is always created?